File: pine-java/src/main/java/page/liam/pine/operators/ReorderShuffle.java
package page.liam.pine.operators;

import page.liam.pine.*;

import java.nio.charset.StandardCharsets;
import java.util.*;

/**
 * Operator: reorder_shuffle_by_salt
 * Metadata contract
 *   CommonInput:  [<salt_fields...>]
 *   CommonOutput: []
 *   ItemInput:    [<item_key_field>]
 *   ItemOutput:   []
 */
public class ReorderShuffle extends AbstractOperator {

    @Override
    public void init(Map<String, Object> params) {}

    @Override
    public void execute(CancellationToken token, OperatorInput input, OperatorOutput output) {
        int n = input.itemCount();
        if (n == 0) return;

        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < commonInput.size(); i++) {
            if (i > 0) sb.append('|');
            sb.append(anyToString(input.common(commonInput.get(i))));
        }
        sb.append('|');
        String saltPrefix = sb.toString();

        String itemField = itemInput.get(0);
        double[] ranks = new double[n];
        long[] ids = new long[n];
        Integer[] indices = new Integer[n];

        for (int i = 0; i < n; i++) {
            String itemVal = anyToString(input.item(i, itemField));
            String key = saltPrefix + itemVal;
            ranks[i] = hashToUnitInterval(key);
            ids[i] = parseUint64(itemVal);
            indices[i] = i;
        }

        Arrays.sort(indices, (a, b) -> {
            int cmp = Double.compare(ranks[a], ranks[b]);
            if (cmp != 0) return cmp;
            return Long.compareUnsigned(ids[a], ids[b]);
        });

        List<Integer> order = new ArrayList<>(n);
        for (int idx : indices) {
            order.add(idx);
        }
        output.setItemOrder(order);
    }

    private static double hashToUnitInterval(String s) {
        long h = fnv64a(s);
        double unsigned = (h & 0x7FFFFFFFFFFFFFFFL) + (h < 0 ? 9223372036854775808.0 : 0.0);
        return unsigned / 18446744073709551616.0;
    }

    private static long fnv64a(String s) {
        byte[] data = s.getBytes(StandardCharsets.UTF_8);
        long hash = 0xcbf29ce484222325L;
        for (byte b : data) {
            hash ^= (b & 0xFF);
            hash *= 0x100000001b3L;
        }
        return hash;
    }

    private static String anyToString(Object v) {
        if (v == null) return "";
        if (v instanceof String) return (String) v;
        if (v instanceof Number) {
            double d = ((Number) v).doubleValue();
            if (d == (long) d && !Double.isInfinite(d)) return Long.toString((long) d);
            return GoFormat.formatG(d);
        }
        return v.toString();
    }

    private static long parseUint64(String s) {
        try {
            return Long.parseUnsignedLong(s);
        } catch (NumberFormatException e) {
            return 0L;
        }
    }
}

File: pine-python/pine/operators/reorder_shuffle.py
from __future__ import annotations

from pine.cancellation import CancellationToken
from pine.go_format import format_g
from pine.operator import (
    AbstractOperator,
    ConsumesRowSet,
    MutatesRowSet,
    OperatorInput,
    OperatorOutput,
    OperatorParams,
)


class ReorderShuffle(AbstractOperator, ConsumesRowSet, MutatesRowSet):
    def __init__(self):
        pass

    def init(self, params: OperatorParams):
        pass

    def execute(
        self, token: CancellationToken, input_: OperatorInput,
        output: OperatorOutput,
    ) -> None:
        n = input_.item_count()
        if n == 0:
            return

        parts: list[str] = []
        for i, field in enumerate(self.common_input()):
            if i > 0:
                parts.append("|")
            parts.append(_any_to_string(input_.common(field)))
        parts.append("|")
        salt_prefix = "".join(parts)

        item_field = self.item_input()[0]
        ranks: list[float] = []
        ids: list[int] = []
        indices = list(range(n))

        for i in range(n):
            item_val = _any_to_string(input_.item(i, item_field))
            key = salt_prefix + item_val
            ranks.append(_hash_to_unit_interval(key))
            ids.append(_parse_uint64(item_val))

        indices.sort(key=lambda idx: (ranks[idx], ids[idx]))
        output.set_item_order(indices)


def _hash_to_unit_interval(s: str) -> float:
    h = _fnv64a(s)
    # Convert signed 64-bit to unsigned
    unsigned = h & 0xFFFFFFFFFFFFFFFF
    return unsigned / 18446744073709551616.0


def _fnv64a(s: str) -> int:
    data = s.encode("utf-8")
    hash_val = 0xCBF29CE484222325
    mask = 0xFFFFFFFFFFFFFFFF
    for b in data:
        hash_val ^= b
        hash_val = (hash_val * 0x100000001B3) & mask
    # Return as signed 64-bit for consistency
    if hash_val >= 0x8000000000000000:
        hash_val -= 0x10000000000000000
    return hash_val


def _any_to_string(v) -> str:
    if v is None:
        return ""
    if isinstance(v, str):
        return v
    if isinstance(v, int):
        return str(v)
    if isinstance(v, float):
        return format_g(v)
    return str(v)


def _parse_uint64(s: str) -> int:
    try:
        val = int(s)
        if val < 0:
            return 0
        return val & 0xFFFFFFFFFFFFFFFF
    except (ValueError, TypeError):
        return 0
